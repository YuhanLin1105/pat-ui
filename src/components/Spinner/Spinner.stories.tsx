import React from 'react';
import Spinner from './Spinner';

export default {
  title: 'Spinner',
  component: Spinner,
};

export const DefaultSpinner = () => <Spinner />;

export const DiffSizeSpinner = () => (
  <div>
    <Spinner size="sm"/>
    <p>Small spinner</p>
    <Spinner />
    <p>Default spinner</p>
    <Spinner size="lg" />
    <p>Large spinner</p>
  </div>
);

export const DiffColorSpinner = () => (
  <div>
    <Spinner color="light" />
    <p>Light spinner</p>
    <Spinner />
    <p>Default (Dark) spinner</p>
  </div>
);