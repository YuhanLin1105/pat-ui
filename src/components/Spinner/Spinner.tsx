import React, { FC, HTMLAttributes } from 'react';
import { classNames } from '../../utils/classNames';

type SpinnerSize = 'sm' | 'md' | 'lg';
type SpinnerColor = 'light' | 'dark';

interface ISpinnerProps {
  // size of the spinner, medium is the default
  size?: SpinnerSize;
  color?: SpinnerColor;
  className?: string;
}

export type SpinnerProps = ISpinnerProps & HTMLAttributes<HTMLDivElement>;

const Spinner: FC<SpinnerProps> = ({ size, color, className }) => {
  let classes = classNames('spinner', {
    [`spinner-${size}`]: !!size,
    [`spinner-${color}`]: !!color,
  });
  if (className) classes += ' ' + className;

  let spinner = (
    <div className={classes} data-testid="spinner-element">
      <div></div>
      <div></div>
      <div></div>
      <div></div>
    </div>
  );

  return spinner;
};

Spinner.defaultProps = {
  size: 'md',
  color: 'dark',
};

export default Spinner;
