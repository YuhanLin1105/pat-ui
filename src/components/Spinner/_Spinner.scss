.spinner {
  display: block;
  position: relative;
  div {
    box-sizing: border-box;
    display: block;
    position: absolute;
    border-radius: 50%;
  }
  div:nth-child(1) {
    animation-delay: -0.45s;
  }
  div:nth-child(2) {
    animation-delay: -0.3s;
  }
  div:nth-child(3) {
    animation-delay: -0.15s;
  }
}

.spinner-md {
  @include spinner-container(50px);
  div {
    @include spinner-size(40px, 6px, $gray-400);
    animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  }
}
.spinner-sm {
  @include spinner-container(40px);
  div {
    @include spinner-size(30px, 5px, $gray-400);
    animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  }
}
.spinner-lg {
  @include spinner-container(60px);
  div {
    @include spinner-size(50px, 7px, $gray-400);
    animation: spinner 1.2s cubic-bezier(0.5, 0, 0.5, 1) infinite;
  }
}

@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
