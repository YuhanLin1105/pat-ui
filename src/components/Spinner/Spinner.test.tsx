import React from 'react';
import { render } from '@testing-library/react';
import Spinner, { SpinnerProps } from './Spinner';

describe('Spinner', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(<Spinner />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with different props', () => {
    const spinnerProps: SpinnerProps = {
      size: 'lg',
      color: 'light',
      className: 'test',
    };
    const { getByTestId } = render(<Spinner {...spinnerProps} />);
    const spinnerElement = getByTestId('spinner-element');
    expect(spinnerElement).toBeInTheDocument();
    expect(spinnerElement).toHaveClass('spinner spinner-lg spinner-light test');
  });
});
