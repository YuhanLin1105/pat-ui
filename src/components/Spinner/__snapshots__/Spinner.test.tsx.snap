// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="spinner spinner-md spinner-dark"
    data-testid="spinner-element"
  >
    <div />
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;
